src/ddmd/dcache.d

```diff
@@ -0,0 +1,377 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (c) 1999-2017 by Digital Mars, All Rights Reserved
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/ddmd/dcache.d, _dcache.d)
+ *
+ * Dcache - a inmplementation if a distributed cache over memory-mapped file.
+ * It doesn't try to be smart and has most basic interface in the way of get/put,
+ * both operations are synchronious and "atomic" by using a simple spin-lock.
+ * 
+ * Intended use
+ *
+ * Allow DMD instances (and frontends on its base) to share CTFE computation results and such,
+ * importantly a key must reflect the state of all dependent entities such as contents of files.
+ *
+ * TODOs:
+ * - eviction policy (e.g. time based expiration)
+ * - sizing policy (is 128 Mb enough for everybody?)
+ * - better hash map algorithm (it's really dumb)
```

Comment:
I'm using a form of cuckoo hashing for memoize.
Takes care of eviction and collision resolution https://github.com/dlang/phobos/blob/27072f8282fd60ef2af025280d217729d2b5770f/std/functional.d#L1044.